service/frontend/nexus_incoming_service_registry.go

```diff
@@ -0,0 +1,406 @@
+		GetNexusIncomingService(context.Context, *operatorservice.GetNexusIncomingServiceRequest) (*operatorservice.GetNexusIncomingServiceResponse, error)
+		ListNexusIncomingServices(context.Context, *operatorservice.ListNexusIncomingServicesRequest) (*operatorservice.ListNexusIncomingServicesResponse, error)
+	}
+
+	registry struct {
+		status           int32
+		serviceDataReady *future.FutureImpl[struct{}]
+
+		sync.RWMutex     // protects tableVersion, services, serviceIDsSorted, and servicesByName
+		tableVersion     int64
+		services         []*nexus.IncomingService // must maintain same sort order as serviceIDsSorted
+		serviceIDsSorted []string                 // parallel array to services of sorted serviceIDs to support pagination during ListNexusIncomingServices
+		servicesByName   map[string]*nexus.IncomingService
+
+		refreshPoller                  *goro.Handle
+		refreshServicesLongPollTimeout dynamicconfig.DurationPropertyFn
+		refreshServicesLongPollMinWait dynamicconfig.DurationPropertyFn
+		refreshServicesRetryPolicy     backoff.RetryPolicy
+
+		matchingClient matchingservice.MatchingServiceClient
+		persistence    p.NexusIncomingServiceManager
+	}
+)
+
+func newNexusIncomingServiceRegistry(
+	listNexusIncomingServicesLongPollTimeout dynamicconfig.DurationPropertyFn,
+	listNexusIncomingServicesLongPollMinWait dynamicconfig.DurationPropertyFn,
+	listNexusIncomingServicesRetryPolicy backoff.RetryPolicy,
+	matchingClient matchingservice.MatchingServiceClient,
+	persistence p.NexusIncomingServiceManager,
+) NexusIncomingServiceRegistry {
+	return &registry{
+		status:                         common.DaemonStatusInitialized,
+		serviceDataReady:               future.NewFuture[struct{}](),
+		refreshServicesLongPollTimeout: listNexusIncomingServicesLongPollTimeout,
+		refreshServicesLongPollMinWait: listNexusIncomingServicesLongPollMinWait,
+		refreshServicesRetryPolicy:     listNexusIncomingServicesRetryPolicy,
+		matchingClient:                 matchingClient,
+		persistence:                    persistence,
+	}
+}
+
+func (r *registry) Start() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	refreshCtx := headers.SetCallerInfo(
+		context.Background(),
+		headers.SystemBackgroundCallerInfo,
+	)
+
+	r.refreshPoller = goro.NewHandle(refreshCtx).Go(r.refreshServices)
+}
+
+func (r *registry) Stop() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	r.refreshPoller.Cancel()
+	<-r.refreshPoller.Done()
+}
+
+func (r *registry) CreateOrUpdateNexusIncomingService(
+	ctx context.Context,
+	request *operatorservice.CreateOrUpdateNexusIncomingServiceRequest,
+) (*operatorservice.CreateOrUpdateNexusIncomingServiceResponse, error) {
+	resp, err := r.matchingClient.CreateOrUpdateNexusIncomingService(ctx, &matchingservice.CreateOrUpdateNexusIncomingServiceRequest{
```

Comment:
Please validate the request.

I'm not sure if this was in the design doc but we should:
1. restrict the number of characters for the service name based on a dynamic config (default to 200 chars)
2. restrict the character types based on the pattern in the latest API commit
3. validate that the namespace it's referencing exists
4. validate the task queue name
5. validate the total size of the metadata map against a dynamic config value (default to 4k)